### ISSUE TITLE
Numpy array alias as type regression between 2.9.2 and 2.10.0 (and later)

### ISSUE THREAD

**Issue by DenSinH** [pydantic/pydantic#11000]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

The below code builds (and correctly prints np.ndarray[100, np.uint8] as output) in version 2.9.2, but raises an exception in version 2.10.0 and later. Perhaps I should be using a different annotation for pytype, or there is an easier way of doing what I want to do (get the generic parameter at runtime), but this way of doing it used to work but fails now. Any suggestions of using a better method to do this are welcome of course.

```
Traceback (most recent call last):
  File "C:\projects\playground\main.py", line 22, in <module>
    x = BaseGeneric[np.ndarray[100, np.uint8]](name="x")
        ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\main.py", line 806, in __class_getitem__
    submodel = _generics.create_generic_submodel(model_name, origin, args, params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generics.py", line 137, in create_generic_submodel
    created_model = meta(
                    ^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_model_construction.py", line 226, in __new__
    complete_model_class(
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_model_construction.py", line 658, in complete_model_class
    schema = cls.__get_pydantic_core_schema__(cls, handler)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\main.py", line 697, in __get_pydantic_core_schema__
    return handler(source)
           ^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_schema_generation_shared.py", line 84, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generate_schema.py", line 612, in generate_schema
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generate_schema.py", line 881, in _generate_schema_inner
    return self._model_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generate_schema.py", line 693, in _model_schema
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generate_schema.py", line 1073, in _generate_md_field_schema
    common_field = self._common_field_schema(name, field_info, decorators)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generate_schema.py", line 1261, in _common_field_schema
    schema = self._apply_annotations(
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generate_schema.py", line 2051, in _apply_annotations
    schema = get_inner_schema(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_schema_generation_shared.py", line 84, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generate_schema.py", line 2126, in new_handler
    schema = metadata_get_schema(source, get_inner_schema)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\functional_validators.py", line 837, in __get_pydantic_core_schema__
    original_schema = handler(source)
                      ^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_schema_generation_shared.py", line 84, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generate_schema.py", line 2032, in inner_handler
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generate_schema.py", line 886, in _generate_schema_inner
    return self.match_type(obj)
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generate_schema.py", line 988, in match_type
    return self._match_generic_type(obj, origin)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generate_schema.py", line 1030, in _match_generic_type
    return self._subclass_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\projects\playground\venv\Lib\site-packages\pydantic\_internal\_generate_schema.py", line 1649, in _subclass_schema
    raise TypeError(f'Expected a class, got {type_param!r}')
TypeError: Expected a class, got numpy.ndarray[100, numpy.uint8]
```

### Example Code

```Python
from pydantic import BaseModel, SkipValidation, Field
import numpy as np
import typing
from typing import Generic, TypeVar, Annotated
import functools

_T = TypeVar("_T")


class BaseGeneric(BaseModel, Generic[_T]):
    pytype: Annotated[type[_T], SkipValidation, Field(frozen=True, exclude=True)]

    def __init__(self, **kwargs):
        super().__init__(pytype=self._cls_get_pytype(), **kwargs)

    @classmethod
    @functools.cache
    def _cls_get_pytype(cls) -> type[_T]:
        return typing.get_args(cls.model_fields["pytype"].annotation)[0]


x = BaseGeneric[np.ndarray[100, np.uint8]](name="x")
print(x.pytype)
```

### Python, Pydantic & OS Version

```Text
Works on: 2.9.2
Fails on: 2.10.0 and 2.10.2 that I tried
```

**Comment by sydney-runkle** [comment]
@DenSinH,

Thanks for the report. Honestly, I don't think this was meant to be supported in earlier versions.

I don't exactly understand what you're going for here - perhaps we could help you figure out another way to get your desired outcome?

**Comment by DenSinH** [comment]
Another way would be perfectly fine, I think the way I am doing it is pretty hacky anyway. Basically, I have a generic model, and I am trying to get the generic parameter at runtime (within the constructor, it need not be stored as a field though). It is supposed to be a type (or a GenericAlias like `np.ndarray[1000, float]`). The way I am currently getting the type parameter is that I add a field to my model, of type `type[_T]`, then at runtime, I read out the annotation of this field and retrieve the type from it, assigning it to the corresponding field.

It doesn't work in the later versions because `np.ndarray[...]` is not a type, but a GenericAlias, and the construction of the submodel fails.

In short, what I am trying to achieve is this: I have a model inheriting from `Generic[_T]`, and I want to determine the value of `_T` at runtime (within the constructor, it need not be a field necessarily though), for a given instance of my model.

**Comment by sydney-runkle** [comment]
> It doesn't work in the later versions because np.ndarray[...] is not a type, but a GenericAlias, and the construction of the submodel fails.

Indeed, this is what's happening now.

Looking into alternative solutions for you. Thanks for the explanation!

**Comment by sydney-runkle** [comment]
How about this?

```py
from pydantic import BaseModel
from typing import Generic, TypeVar

_T = TypeVar("_T")


class BaseGeneric(BaseModel, Generic[_T]):
    field: _T

x = BaseGeneric[str](field="x")
print(x.__pydantic_generic_metadata__)
#> {'origin': <class '__main__.BaseGeneric'>, 'args': (<class 'str'>,), 'parameters': ()}
```

**Comment by sydney-runkle** [comment]
Going to close this as resolved for now, as I don't think we're going to change the `inspect.isclass` check that you were running into. Happy to continue discussing a solution to your use case, though!

**Comment by DenSinH** [comment]
This looks like it might work, though it may break with inheritance. For example, I might inherit from `BaseGeneric[bool]` explicitly, and the metadata will not contain any generic parameter. I'm not currently at my laptop so I can't try it but I think that would break, and I do need it for my use case.

**Comment by sydney-runkle** [comment]
cc @dmontagu, you might have a better idea here

**Comment by sydney-runkle** [comment]
Reopening, I think we might reconsider greater support for `type[SomeGenericAlias]`

**Comment by DenSinH** [comment]
I was just messing around with it again, something like this does what I want, but ideally, I do not want to edit the `model_config` (or even better: not have the `pytype` field at all)
```python
from typing import Generic, TypeVar, Annotated
from pydantic import BaseModel, Field, ConfigDict, SkipValidation
import numpy as np

_T = TypeVar("_T")


class BaseGeneric(BaseModel, Generic[_T]):
    model_config = ConfigDict(
        arbitrary_types_allowed=True,
    )

    pytype: Annotated[_T, SkipValidation, Field(default=None, validate_default=False, exclude=True, repr=False, frozen=True)]

    def get_type(self) -> type[_T]:
        return self.model_fields["pytype"].annotation


class BoolDerived(BaseGeneric[bool]):
    pass


test = BoolDerived()
x = BaseGeneric[np.ndarray[1000, np.uint8]]()
print(test.get_type())
print(x.get_type())
```

**Comment by Viicos** [comment]
@DenSinH, I think using [private attributes](https://docs.pydantic.dev/dev/concepts/models/#private-model-attributes) will give you more flexibility:

```python
from pydantic import BaseModel
import numpy as np
from typing import Generic, TypeVar, Annotated, Any, get_args

_T = TypeVar("_T")


class BaseGeneric(BaseModel, Generic[_T]):
    _pytype: type[_T]

    def model_post_init(self, context: Any) -> None:
        self._pytype = type(self).__pydantic_generic_metadata__['args'][0]


x = BaseGeneric[np.ndarray[100, np.uint8]](name="x")
print(x._pytype)
```

A Pydantic design constraint is to have private attributes named with a leading underscore. You can alternatively provide a `pytype` property to expose the private attribute.

---

Whether we should allow the following or not is debatable:

```python
class Model[T](BaseModel):
    f: type[T]

param: typing._GenericAlias  # e.g. List[int]

Model[param]  # currently fails
```

If we allow this, users will except things like this to be validated by Pydantic:

```python
T_co = TypeVar('T', covariant=True)

class GenericClass(Generic[T_Co]): ...

class Model(BaseModel):
    a: type[GenericClass[int]]

class SubInt(int): ...

class SubGenericClass(GenericClass[SubInt]): ...

Model(a=SubGenericClass)  # OK; # Not OK if T_co was invariant or contravariant
```

So we would need to introduce the same mechanisms mentioned in https://github.com/pydantic/pydantic/issues/10462.

**Comment by Viicos** [comment]
Here is an alternative implementation:

```python
from pydantic import BaseModel
from pydantic_core import PydanticUndefined
import numpy as np
from typing import ClassVar, Generic, TypeVar, Any


_T = TypeVar("_T")


class BaseGeneric(BaseModel, Generic[_T]):
    # Or, might be more suitable: ClassVar[TypeForm[_T]]  (see PEP 747)
    pytype: ClassVar[type[_T]] = PydanticUndefined  # type: ignore (see , see https://github.com/python/typing/discussions/1424)

    @classmethod
    def __pydantic_init_subclass__(cls, **kwargs: Any) -> None:
        if cls.pytype is PydanticUndefined:
            args = cls.__pydantic_generic_metadata__['args']
            if args:
                cls.pytype = args[0]


x = BaseGeneric[np.ndarray[100, np.uint8]]()
print(x.pytype)

class BoolDerived(BaseGeneric[bool]):
    pass

test = BoolDerived()
print(test.pytype)
```

It makes more sense for `pytype` to be a class variable, as as far as type checkers are concerned, `BaseGeneric[<typ>]` is still a class.

Regarding my second point in https://github.com/pydantic/pydantic/issues/11000#issuecomment-2527730758, I'll open a PR to provide a better error message.
